### ISSUE TITLE
please check the code of "discrete == True" in model_search.py

### ISSUE THREAD

**Issue by nabk89** [automl/RobustDARTS#7]
Hi, I appreciate for sharing your work including the paper and this open-source code.

I have a question about measuring "performance drop" in 4.2 section of the paper (figure 5(b)).

Let me explain my concern with an example. My understanding of RobustDARTS is as follows:
  1) validation accuracy of models are measured after search phase (i.e., after 50 epochs of search)
  2) weighting in mixed operation (i.e., softmax output of alpha) is not one-hot before discretizing alpha, and is converted into one-hot vector after discretizing to leave one operation per edge.
      ex) [one edge] alpha after 50 epoch search = [0.1 0.4 0.2 0.1] 
                              --> softmax(alpha) = [0.2245 0.3030 0.2481 0.2245]
                              --> this is used for weighting of mixed op before discretizing.
        * after discretizing --> the weighting of mixed op is converted into an one-hot vector: [0 1 0 0].

Please check if my understanding is right.

If it is, the code of "discrete == True" in forward() of Network class in model_search.py seems to have an error.
In your code, weighting is [0.1 0.4 0.2 0.1] if discrete (without softmax activation), otherwise [0.2245 0.3030 0.2481 0.2245] (with softmax activation).
In my opinion, if discrete, the weighting should be [0 1 0 0].

Please check this issue.
If I am misunderstanding the discretizing process, I would like to ask for a clearer explanation of the discretizing step and the measurement of "performance drop."

Thank you.

**Comment by arberzela** [comment]
Thank you very much for reading our paper! You are of course right about that, sorry for the confusion!
We use the `discrete=True` only when running Random Search with Weight Sharing: https://github.com/automl/RobustDARTS/blob/master/src/search/randomNAS/darts_wrapper_discrete.py#L111
There the set of alphas is [0 1 0 0] and I added that condition in the forward pass to avoid applying a softmax to the discrete alphas.

For that experiments in section 4.2 I used another script in a post-hoc manner. After loading the saved search model in `train_search.py` you can use this piece of code:
```
genotype, _ = model.genotype()
logging.info('Before loading genotype = %s', genotype)
utils.load_checkpoint(model, args.model_path, 1)
valid_acc_1, valid_obj_1 = infer(valid_queue, model, criterion)
logging.info('VALID ACC: %f', valid_acc_1)
genotype, _ = model.genotype()
logging.info('After loading genotype = %s', genotype)
print(F.softmax(model.alphas_normal, dim=-1))
print(F.softmax(model.alphas_reduce, dim=-1))
alphas_normal = utils.get_one_hot(model.alphas_normal)
alphas_reduce = utils.get_one_hot(model.alphas_reduce)
print(alphas_normal)
print(alphas_reduce)
model.alphas_normal = Variable(alphas_normal).cuda()
model.alphas_reduce = Variable(alphas_reduce).cuda()
valid_acc_2, valid_obj_2 = infer(valid_queue, model, criterion)
logging.info('ONE-HOT VALID ACC: %f', valid_acc_2)
logging.info('ACC drop: %f  LOSS drop: %f', valid_acc_1 - valid_acc_2, valid_obj_1 - valid_obj_2)
```

I also just pushed the `get_one_hot` function to `utils.py`

Hope this helps!
Cheers,
Arber
